File: backend/src/main/java/com/backend/domain/repository/service/mapper/CicdInfoMapper.java
package com.backend.domain.repository.service.mapper;

import com.backend.domain.repository.dto.response.RepositoryData;
import com.backend.domain.repository.dto.response.github.TreeResponse;
import org.springframework.stereotype.Component;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

@Component
public class CicdInfoMapper {
    // ResponseData CI/CD 관련 [CI/CD 존재 여부 관련]
    private static final String GITHUB_WORKFLOWS_PATTERN = "^\\.github/workflows/.*\\.(yml|yaml)$";

    private static final List<String> CICD_FILE_PATTERNS = List.of(
            // 주요 CI/CD 플랫폼
            "^\\.gitlab-ci\\.(yml|yaml)$",
            "^\\.travis\\.(yml|yaml)$",
            "^azure-pipelines\\.(yml|yaml)$",
            "^\\.circleci/config\\.(yml|yaml)$",
            "^\\.github/actions/.*\\.(yml|yaml)$",
            "^\\.bitbucket-pipelines\\.(yml|yaml)$",
            "^\\.appveyor\\.(yml|yaml)$"
    );

    private static final List<String> JENKINSFILE_PATTERNS = List.of(
            ".*/[Jj]enkinsfile$",
            "^[Jj]enkinsfile$",
            ".*/[Jj]enkinsfile\\..+$",
            "^[Jj]enkinsfile\\..+$"
    );

    private static final List<String> BUILD_FILE_PATTERNS = List.of(
            // Java 빌드 도구
            ".*pom\\.xml$",
            ".*build\\.gradle(\\.kts)?$",
            ".*gradlew(\\.bat)?$",
            ".*mvnw(\\.cmd)?$",

            // JavaScript/Node.js
            ".*package\\.json$",
            ".*(yarn\\.lock|pnpm-lock\\.yaml|package-lock\\.json)$",

            // Python
            ".*requirements\\.txt$",
            ".*setup\\.py$",
            ".*pyproject\\.toml$",
            ".*setup\\.cfg$",

            // 기타 주요 언어
            ".*Cargo\\.toml$",
            ".*go\\.mod$",
            ".*Gemfile(\\.lock)?$",
            ".*composer\\.(json|lock)$",

            // C/C++ 및 범용
            ".*CMakeLists\\.txt$",
            ".*Makefile$",
            ".*build\\.xml$"
    );

    private static final List<String> DOCKERFILE_PATTERNS = List.of(
            ".*(?i)dockerfile.*$",
            ".*(?i)docker-compose.*\\.(yml|yaml)$",
            ".*(?i)compose.*\\.(yml|yaml)$"
    );

    public void mapCicdInfo(RepositoryData data, TreeResponse response) {
        if (response == null || response.tree() == null || response.tree().isEmpty()) {
            setEmptyCicdData(data);
            return;
        }

        List<String> filePaths = extractFilePaths(response);

        // CI/CD 설정 확인
        List<String> cicdFiles = findCicdFiles(filePaths);
        data.setHasCICD(!cicdFiles.isEmpty());
        data.setCicdFiles(cicdFiles);

        // 빌드 스크립트 확인
        List<String> buildFiles = findBuildFiles(filePaths);
        data.setHasBuildFile(!buildFiles.isEmpty());
        data.setBuildFiles(buildFiles);

        // Dockerfile 별도 확인
        boolean hasDockerfile = hasDockerFiles(filePaths);
        data.setHasDockerfile(hasDockerfile);
    }

    private void setEmptyCicdData(RepositoryData data) {
        data.setHasCICD(false);
        data.setCicdFiles(Collections.emptyList());
        data.setHasDockerfile(false);
    }

    private List<String> extractFilePaths(TreeResponse response) {
        if (response == null || response.tree() == null) {
            return Collections.emptyList();
        }

        return response.tree().stream()
                .filter(item -> "blob".equals(item.type()))
                .map(TreeResponse.TreeItem::path)
                .collect(Collectors.toList());
    }

    private List<String> findCicdFiles(List<String> filePaths) {
        return filePaths.stream()
                .filter(this::isCicdFile)
                .collect(Collectors.toList());
    }

    private boolean isCicdFile(String filePath) {
        if (filePath.matches(GITHUB_WORKFLOWS_PATTERN)) {
            return true;
        }

        if (JENKINSFILE_PATTERNS.stream().anyMatch(filePath::matches)) {
            return true;
        }

        return CICD_FILE_PATTERNS.stream()
                .anyMatch(filePath::matches);
    }

    private List<String> findBuildFiles(List<String> filePaths) {
        return filePaths.stream()
                .filter(this::isBuildFile)
                .collect(Collectors.toList());
    }

    private boolean isBuildFile(String filePath) {
        return BUILD_FILE_PATTERNS.stream()
                .anyMatch(filePath::matches);
    }

    private boolean hasDockerFiles(List<String> filePaths) {
        return filePaths.stream()
                .anyMatch(path -> DOCKERFILE_PATTERNS.stream()
                        .anyMatch(path::matches));
    }
}

File: backend/src/main/java/com/backend/domain/repository/service/mapper/CicdInfoMapper.kt
package com.backend.domain.repository.service.mapper

import com.backend.domain.repository.dto.response.RepositoryData
import com.backend.domain.repository.dto.response.github.TreeResponse
import org.springframework.stereotype.Component

@Component
class CicdInfoMapper {

    fun mapCicdInfo(data: RepositoryData, response: TreeResponse?) {
        if (response == null || response.tree.isNullOrEmpty()) {
            setEmptyCicdData(data)
            return
        }

        val filePaths = extractFilePaths(response)

        // CI/CD 파일
        val cicdFiles = findCicdFiles(filePaths)
        data.hasCICD = cicdFiles.isNotEmpty()
        data.cicdFiles = cicdFiles

        // Build 파일
        val buildFiles = findBuildFiles(filePaths)
        data.hasBuildFile = buildFiles.isNotEmpty()
        data.buildFiles = buildFiles

        // Dockerfile
        data.hasDockerfile = hasDockerFiles(filePaths)
    }

    private fun setEmptyCicdData(data: RepositoryData) {
        data.hasCICD = false
        data.cicdFiles = emptyList()
        data.hasDockerfile = false
    }

    private fun extractFilePaths(response: TreeResponse): List<String> {
        return response.tree
            ?.filter { it.type == "blob" }
            ?.mapNotNull { it.path }
            ?: emptyList()
    }

    private fun findCicdFiles(filePaths: List<String>): List<String> {
        return filePaths.filter { isCicdFile(it) }
    }

    private fun isCicdFile(filePath: String): Boolean {
        if (filePath.matches(GITHUB_WORKFLOWS_PATTERN.toRegex())) return true
        if (JENKINSFILE_PATTERNS.any { filePath.matches(it.toRegex()) }) return true
        return CICD_FILE_PATTERNS.any { filePath.matches(it.toRegex()) }
    }

    private fun findBuildFiles(filePaths: List<String>): List<String> {
        return filePaths.filter { isBuildFile(it) }
    }

    private fun isBuildFile(filePath: String): Boolean {
        return BUILD_FILE_PATTERNS.any { filePath.matches(it.toRegex()) }
    }

    private fun hasDockerFiles(filePaths: List<String>): Boolean {
        return filePaths.any { path ->
            DOCKERFILE_PATTERNS.any { regex -> path.matches(regex.toRegex()) }
        }
    }

    companion object {
        private const val GITHUB_WORKFLOWS_PATTERN = "^\\.github/workflows/.*\\.(yml|yaml)$"

        private val CICD_FILE_PATTERNS = listOf(
            "^\\.gitlab-ci\\.(yml|yaml)$",
            "^\\.travis\\.(yml|yaml)$",
            "^azure-pipelines\\.(yml|yaml)$",
            "^\\.circleci/config\\.(yml|yaml)$",
            "^\\.github/actions/.*\\.(yml|yaml)$",
            "^\\.bitbucket-pipelines\\.(yml|yaml)$",
            "^\\.appveyor\\.(yml|yaml)$"
        )

        private val JENKINSFILE_PATTERNS = listOf(
            ".*/[Jj]enkinsfile$",
            "^[Jj]enkinsfile$",
            ".*/[Jj]enkinsfile\\..+$",
            "^[Jj]enkinsfile\\..+$"
        )

        private val BUILD_FILE_PATTERNS = listOf(
            ".*pom\\.xml$",
            ".*build\\.gradle(\\.kts)?$",
            ".*gradlew(\\.bat)?$",
            ".*mvnw(\\.cmd)?$",
            ".*package\\.json$",
            ".*(yarn\\.lock|pnpm-lock\\.yaml|package-lock\\.json)$",
            ".*requirements\\.txt$",
            ".*setup\\.py$",
            ".*pyproject\\.toml$",
            ".*setup\\.cfg$",
            ".*Cargo\\.toml$",
            ".*go\\.mod$",
            ".*Gemfile(\\.lock)?$",
            ".*composer\\.(json|lock)$",
            ".*CMakeLists\\.txt$",
            ".*Makefile$",
            ".*build\\.xml$"
        )

        private val DOCKERFILE_PATTERNS = listOf(
            ".*(?i)dockerfile.*$",
            ".*(?i)docker-compose.*\\.(yml|yaml)$",
            ".*(?i)compose.*\\.(yml|yaml)$"
        )
    }
}
